File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractQueryCommand.java
package io.codemodder.plugins.maven.operator;

import io.github.pixee.security.BoundedLineReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Common Base Class - Meant to be used by Simple Queries using either Invoker and/or Embedder, thus
 * relying on dependency:tree mojo outputting into a text file - which might be cached.
 */
abstract class AbstractQueryCommand extends AbstractCommand {

  protected static final Logger LOGGER = LoggerFactory.getLogger(AbstractQueryCommand.class);

  /**
   * Generates a temporary file path used to store the output of the
   *
   * <pre>dependency:tree</pre>
   *
   * mojo
   *
   * @param pomFilePath POM Original File Path
   */
  private Path getOutputPath(Path pomFilePath) {
    Path basePath = pomFilePath.getParent();

    String outputBasename = String.format("output-%08X.txt", pomFilePath.hashCode());

    return basePath.resolve(outputBasename);
  }

  /**
   * Given a POM URI, returns a File Object
   *
   * @param d POMDocument
   */
  private Path getPomFilePath(POMDocument d) throws URISyntaxException {
    return Paths.get(d.getPomPath().toURI());
  }

  /**
   * Abstract Method to extract dependencies
   *
   * @param outputPath Output Path to where to store the content
   * @param pomFilePath Input Pom Path
   * @param c Project Model
   */
  protected abstract void extractDependencyTree(Path outputPath, Path pomFilePath, ProjectModel c);

  /**
   * Internal Holder Variable
   *
   * <p>Todo: OF COURSE IT BREAKS THE PROTOCOL
   */
  protected Collection<Dependency> result = null;

  /**
   * Retrieves the collection of dependencies resulting from the query command execution.
   *
   * @return A collection of Dependency objects representing project dependencies.
   */
  public Collection<Dependency> getResult() {
    return result;
  }

  /**
   * We declare the main logic here - details are made in the child classes for now Executes the
   * query command, extracting and processing project dependencies.
   *
   * @param pm The ProjectModel representing the project on which to execute the command.
   * @return true if the execution was successful, false otherwise.
   * @throws URISyntaxException If there is an issue with URIs in the execution.
   * @throws IOException If an IO error occurs during execution.
   */
  @Override
  public boolean execute(ProjectModel pm) throws URISyntaxException, IOException {
    Path pomFilePath = getPomFilePath(pm.getPomFile());
    Path outputPath = getOutputPath(pomFilePath);

    if (Files.exists(outputPath)) {
      Files.delete(outputPath);
    }

    try {
      extractDependencyTree(outputPath, pomFilePath, pm);
    } catch (InvalidContextException e) {
      return false;
    }

    result = extractDependencies(outputPath).values();

    return true;
  }

  /**
   * Given a File containing the output of the dependency:tree mojo, read its contents and parse,
   * creating an array of dependencies
   *
   * <p>About the file contents: We receive something such as this, then filter it out:
   *
   * <pre>
   *     br.com.ingenieux:pom-operator:jar:0.0.1-SNAPSHOT
   *     +- xerces:xercesImpl:jar:2.12.1:compile
   *     |  \- xml-apis:xml-apis:jar:1.4.01:compile
   *     \- org.jetbrains.kotlin:kotlin-test:jar:1.5.31:test
   * </pre>
   *
   * @param outputPath file to read
   */
  private Map<String, Dependency> extractDependencies(Path outputPath) throws IOException {
    Map<String, Dependency> dependencyMap = new HashMap<>();
    try (BufferedReader reader = Files.newBufferedReader(outputPath)) {
      String line;
      boolean skipFirstLine = true;
      while ((line = BoundedLineReader.readLine(reader, 5_000_000)) != null) {
        if (skipFirstLine) {
          skipFirstLine = false;
          continue;
        }
        String trimmedLine = trimSpecialChars(line);
        String[] elements = trimmedLine.split(":");

        if (elements.length >= 5) {
          String groupId = elements[0];
          String artifactId = elements[1];
          String packaging = elements[2];
          String version = elements[3];
          String scope = elements[4];

          Dependency dependency =
              new Dependency(groupId, artifactId, version, null, packaging, scope);
          dependencyMap.put(line, dependency);
        }
      }
    }
    return dependencyMap;
  }

  private String trimSpecialChars(String input) {
    char[] specialChars = "+-|\\ ".toCharArray();
    int start = 0;
    int end = input.length();

    while (start < end && isSpecialChar(input.charAt(start))) {
      start++;
    }

    while (end > start && isSpecialChar(input.charAt(end - 1))) {
      end--;
    }

    if (start > 0 || end < input.length()) {
      return input.substring(start, end);
    }

    return input;
  }

  private boolean isSpecialChar(char c) {
    char[] specialChars = "+-|\\ ".toCharArray();
    for (char specialChar : specialChars) {
      if (c == specialChar) {
        return true;
      }
    }
    return false;
  }

  @Override
  public boolean postProcess(ProjectModel c) {
    return false;
  }
}

File: src/main/java/io/github/pixee/maven/operator/AbstractQueryCommand.kt
package io.github.pixee.maven.operator

import io.github.pixee.maven.operator.Util.which
import org.apache.commons.lang3.SystemUtils
import org.apache.maven.shared.invoker.DefaultInvocationRequest
import org.apache.maven.shared.invoker.InvocationRequest
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.nio.file.Paths
import java.util.*

/**
 * Common Base Class - Meant to be used by Simple Queries using either Invoker and/or Embedder, thus
 * relying on dependency:tree mojo outputting into a text file - which might be cached.
 *
 */
abstract class AbstractQueryCommand : AbstractCommand() {
    /**
     * Generates a temporary file path used to store the output of the <pre>dependency:tree</pre> mojo
     *
     * @param pomFilePath POM Original File Path
     */
    private fun getOutputPath(pomFilePath: File): File {
        val basePath = pomFilePath.parentFile

        val outputBasename = "output-%08X.txt".format(pomFilePath.hashCode())

        val outputPath = File(basePath, outputBasename)

        return outputPath
    }

    /**
     * Given a POM URI, returns a File Object
     *
     * @param d POMDocument
     */
    protected fun getPomFilePath(d: POMDocument): File = Paths.get(d.pomPath!!.toURI()).toFile()

    /**
     * Abstract Method to extract dependencies
     *
     * @param outputPath Output Path to where to store the content
     * @param pomFilePath Input Pom Path
     * @param c Project Model
     */
    abstract fun extractDependencyTree(outputPath: File, pomFilePath: File, c: ProjectModel)

    /**
     * Internal Holder Variable
     *
     * Todo: OF COURSE IT BREAKS THE PROTOCOL
     */
    internal var result: Collection<Dependency>? = null

    /**
     * We declare the main logic here - details are made in the child classes for now
     */

    override fun execute(pm: ProjectModel): Boolean {
        val pomFilePath = getPomFilePath(pm.pomFile)

        val outputPath = getOutputPath(pomFilePath)

        if (outputPath.exists()) {
            outputPath.delete()
        }

        try {
            extractDependencyTree(outputPath, pomFilePath, pm)
        } catch (e: InvalidContextException) {
            return false
        }

        this.result = extractDependencies(outputPath).values

        return true
    }

    /**
     * Given a File containing the output of the dependency:tree mojo, read its contents and parse, creating an array of dependencies
     *
     * About the file contents: We receive something such as this, then filter it out:
     *
     * <pre>
     *     br.com.ingenieux:pom-operator:jar:0.0.1-SNAPSHOT
     *     +- xerces:xercesImpl:jar:2.12.1:compile
     *     |  \- xml-apis:xml-apis:jar:1.4.01:compile
     *     \- org.jetbrains.kotlin:kotlin-test:jar:1.5.31:test
     * </pre>
     *
     * @param outputPath file to read
     */
    protected fun extractDependencies(outputPath: File) = outputPath.readLines().drop(1).map {
        it.trim(*"+-|\\ ".toCharArray())
    }.map {
        it to it.split(':')
    }.associate { (line, elements) ->
        val (groupId, artifactId, packaging, version, scope) = elements

        line to Dependency(
            groupId = groupId,
            artifactId = artifactId,
            version = version,
            packaging = packaging,
            scope = scope
        )
    }

    protected fun buildInvocationRequest(
        outputPath: File,
        pomFilePath: File,
        c: ProjectModel
    ): InvocationRequest {
        val props = Properties(System.getProperties()).apply {
            setProperty("outputFile", outputPath.absolutePath)

            val localRepositoryPath = getLocalRepositoryPath(c).absolutePath

            setProperty("maven.repo.local", localRepositoryPath)
        }

        val request: InvocationRequest = DefaultInvocationRequest().apply {
            findMaven(this)

            pomFile = pomFilePath

            isShellEnvironmentInherited = true

            isNoTransferProgress = true
            isBatchMode = true
            isRecursive = false
            profiles = c.activeProfiles.toList()
            isDebug = true

            isOffline = c.offline

            properties = props

            goals = listOf(DEPENDENCY_TREE_MOJO_REFERENCE)
        }

        return request
    }

    /**
     * Locates where Maven is at - HOME var and main launcher script.
     *
     * @param invocationRequest InvocationRequest to be filled up
     */
    private fun findMaven(invocationRequest: InvocationRequest) {
        /*
         * Step 1: Locate Maven Home
         */
        val m2homeEnvVar = System.getenv("M2_HOME")

        if (null != m2homeEnvVar) {
            val m2HomeDir = File(m2homeEnvVar)

            if (m2HomeDir.isDirectory)
                invocationRequest.mavenHome = m2HomeDir
        }

        /**
         * Step 1.1: Try to guess if thats the case
         */
        if (invocationRequest.mavenHome == null) {
            val inferredHome = File(SystemUtils.getUserHome(), ".m2")

            if (!(inferredHome.exists() && inferredHome.isDirectory)) {
                LOGGER.warn(
                    "Inferred User Home - which does not exist or not a directory: {}",
                    inferredHome
                )
            }

            invocationRequest.mavenHome = inferredHome
        }

        /**
         * Step 2: Find Maven Executable given the operating system and PATH variable contents
         */
        val foundExecutable = listOf("mvn", "mvnw").map { which(it) }.firstOrNull()

        if (null != foundExecutable) {
            invocationRequest.mavenExecutable = foundExecutable

            return
        }

        throw IllegalStateException("Missing Maven Home / Executable")
    }

    companion object {
        /**
         * Mojo Reference
         */
        const val DEPENDENCY_TREE_MOJO_REFERENCE =
            "org.apache.maven.plugins:maven-dependency-plugin:3.3.0:tree"

        val LOGGER: Logger = LoggerFactory.getLogger(AbstractQueryCommand::class.java)
    }

    override fun postProcess(c: ProjectModel): Boolean = false
}
